File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.py
class Solution:
    def mostPoints(self, questions: List[List[int]]) -> int:
        def dp(i):
            nonlocal memo
            if i in memo: return memo[i]
            if i >= len(questions): return 0
            solve_cur = questions[i][0] + dp(i + questions[i][1] + 1)
            skip_cur = dp(i + 1)
            memo[i] = max(solve_cur, skip_cur)
            return memo[i]
        memo = {}
        return dp(0)
        

File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.java
class Solution {
    public long mostPoints(int[][] questions) {
        int n = questions.length;
        long [] dp = new long[n + 1];
        
        for (int i = n - 1; i >= 0; i--){
            int pt = questions[i][0], jmp = questions[i][1];
            dp[i] = Math.max(dp[i + 1] , dp[Math.min(i + jmp + 1, n)] + pt);
        }
        return dp[0];
    }
}